### PR TITLE
VZ-6649.  Utility lib for creating a local CAPI management bootstrap cluster

### PR DESCRIPTION
**Pull request by michael-cico**
This represents in-progress changes for Epic VZ-6793, OCNE in a local container runtime.

These are the initial changes to support creating a local CAPI management cluster using KIND with the VZ CLI.  There will be follow-up work to integrate it with a CLI command
- Updated dependencies, 3rdparty license file for KIND 0.14 and CAPI 1.2.0 usage
- A small utility library to support creating a local bootstrap/management cluster with KIND
- Updated generated clients due to updated versions of k8s libraries
- Updated `controller-gen` target to use `go install` to install binary instead of deprecated `go get`

### CODE REVIEW

File: make/generate.mk [hunk 1]
```diff
@@ -44,7 +44,7 @@ generate: controller-gen
 .PHONY: controller-gen
 controller-gen:
 ifeq (, $(shell command -v controller-gen))
-	$(GO) get sigs.k8s.io/controller-tools/cmd/controller-gen@${CONTROLLER_GEN_VERSION}
+	$(GO) install sigs.k8s.io/controller-tools/cmd/controller-gen@${CONTROLLER_GEN_VERSION}
```
Comment (pfmackin) [inline make/generate.mk:47]: why is this changed?

File: pkg/capi/kind_cluster_manager.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	defer os2.RemoveTempFiles(zap.S(), config.Name())
+	capiclient, err := capiInitFunc("") // TODO: do we need to provide a CAPI config?
+	if err != nil {
+		return err
+	}
+	_, err = capiclient.Init(clusterapi.InitOptions{
+		Kubeconfig: clusterapi.Kubeconfig{
+			Path: config.Name(),
+		},
+		InfrastructureProviders: []string{"docker"},
```
Comment (pfmackin) [inline pkg/capi/kind_cluster_manager.go:36]: can you make this a constant.  Also should it be injected or will it always be docker?

File: pkg/capi/kind_cluster_manager.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+	})
+	return err
```
Comment (pfmackin) [inline pkg/capi/kind_cluster_manager.go:38]: We should log error here

File: pkg/capi/kind_cluster_manager.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+	config ClusterConfig
+}
+
+func (r *kindClusterManager) GetKubeConfig() (string, error) {
+	return bootstrapProviderImpl.GetKubeconfig(r.config.ClusterName())
+}
+
+func (r *kindClusterManager) Init() error {
+	config, err := r.createKubeConfigFile()
+	if err != nil {
```
Comment (pfmackin) [inline pkg/capi/kind_cluster_manager.go:24]: We should log error here

File: pkg/capi/kind_cluster_manager.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+		return err
+	}
+	defer os2.RemoveTempFiles(zap.S(), config.Name())
+	capiclient, err := capiInitFunc("") // TODO: do we need to provide a CAPI config?
+	if err != nil {
+		return err
```
Comment (pfmackin) [inline pkg/capi/kind_cluster_manager.go:30]: We should log error here

File: pkg/capi/kind_cluster_manager.go [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+	os2 "github.com/verrazzano/verrazzano/pkg/os"
+	"go.uber.org/zap"
+	"io/ioutil"
+	"os"
+	clusterapi "sigs.k8s.io/cluster-api/cmd/clusterctl/client"
+)
+
+// kindClusterManager ClusterLifecycleManager impl for a KIND-based bootstrap cluster
+type kindClusterManager struct {
+	config ClusterConfig
```
Comment (pfmackin) [inline pkg/capi/kind_cluster_manager.go:15]: where is this set?

File: pkg/capi/cluster_lifecycle.go [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+	Type() string
+	ContainerImage() string
+}
+
+// ClusterLifeCycleManager defines the lifecycle operations of a cluster
+type ClusterLifeCycleManager interface {
+	GetConfig() ClusterConfig
+	GetKubeConfig() (string, error)
+	Create() error
+	Init() error
```
Comment (pfmackin) [inline pkg/capi/cluster_lifecycle.go:35]: Does Init have to be called before each create?

File: pkg/capi/capi_test.go [hunk 8]
```diff
@@ -0,0 +1,107 @@
+// Copyright (c) 2022, Oracle and/or its affiliates.
```
Comment (pfmackin) [inline pkg/capi/capi_test.go:1]: there should be a capi.go file that corresponds to capi_test.go

**Review by pfmackin — APPROVED** [review]

File: make/generate.mk (hunk 1 above)
Comment (michael-cico) [reply make/generate.mk:47]: I was getting a warning from `go get` that using `get` for binaries is deprecated, and that `go install` is preferred.

File: pkg/capi/kind_cluster_manager.go (hunk 2 above)
Comment (michael-cico) [reply pkg/capi/kind_cluster_manager.go:36]: At present, I'm just using that, but we can allow the caller to specify the provider list.  We'll have to for the CLI work that is currently spec'd.

File: pkg/capi/kind_cluster_manager.go (hunk 3 above)
Comment (michael-cico) [reply pkg/capi/kind_cluster_manager.go:38]: It's probably better to let the caller do the logging right?  since this is a util lib

File: pkg/capi/kind_cluster_manager.go (hunk 4 above)
Comment (michael-cico) [reply pkg/capi/kind_cluster_manager.go:24]: Same comment as above, for a util lib I thought it was better leave that to the caller.

File: pkg/capi/cluster_lifecycle.go (hunk 7 above)
Comment (michael-cico) [reply pkg/capi/cluster_lifecycle.go:35]: Init can probably be consolidated with `Create`... wasn't sure if wanted this as a separate step or not.

File: pkg/capi/capi_test.go (hunk 8 above)
Comment (michael-cico) [reply pkg/capi/capi_test.go:1]: This is strictly to contain shared test code, the stubs etc.  I can move that stuff back into one of the other test files if that helps.

**Review by pfmackin — APPROVED** [review]

**Review by desagar — APPROVED** [review]
